### ISSUE TITLE
Trouble creating a re-posh/datascript DB via an AJAX call

### ISSUE THREAD

**Issue by zachcp** [denistakeda/re-posh#17]
Hi @denistakeda,

Thank you for your excellent library.  I've enjoyed playing with it and I have been trying to make an app where I allow users to upload a file which I parse server-side, create a datascript DB, and then  send the full DB back to the browser where I can write components against the DB that allow for data exploration. 

However, I am running into issues connecting the AJAX call to re-posh.  While it seems that my issue is probably due to incorrectly parsing the returned db (and therefore with how I am handling the AJAX call with transit), I think this issue may be relevant to other potential users of this library so I'd like to post it here. If you can help me solve this issue, I would be happy to document the use case as an example in the repo.

Here is the relevant code and the error messages.

```clojure
;; client code
;; In this case I am not POSTing/uploading the file
;; I am simply using a GET call to try to pull a datascript DB from the server
(rp/reg-event-ds
  :upload
  (fn [db [_ ds]]
    (GET
      "/upload"
      :format :transit
      :response-format :transit
      :handler #(re-frame/dispatch  [:create-new-db  %1])
      :error-handler #(js/alert (str "Didn't Work !" " Response:" ds)))
    ds))    ;; pure handlers must return a db (unchanged in this case)

;; create the new ds
(rp/reg-event-ds
  :create-new-db
  (fn  [_ [_  ds]]
    (print-str ds)
    (dt/read-transit-str ds)))

;; server code
;; the upload currently serves and existing datascript db (no upload/parsing is done here)
(defroutes home-routes
  (GET "/" []
       (home-page))
  (GET "/upload" []
    (-> (response/ok genbank/ds1) ;; am I wrapping the ring/transit response incorrectly here?
        (response/header "Content-Type" "application/transit+json"))))
``` 

This image shows the datascript DB is being fetched:

<img width="426" alt="screen shot 2017-12-29 at 2 20 31 pm" src="https://user-images.githubusercontent.com/1260667/34445704-8a8e6f62-eca3-11e7-87c0-4745f77c7e0d.png">

However, I am running into issues parsing the JSON:
<img width="1102" alt="screen shot 2017-12-29 at 2 22 16 pm" src="https://user-images.githubusercontent.com/1260667/34445737-bd21d91e-eca3-11e7-9d94-795f058f2969.png">

So it seems I am having trouble reading and returning the Datascript DB from the server but have been  getting nowhere with my normal approach of fiddling about until things work - in other words, I am clearly missing something. I would  appreciate any pointers you might be able to provide.

Thanks,
zach cp

**Comment by denistakeda** [comment]
Hi @zachcp 
As far as I can see you are using the `re-posh/reg-event-ds` wrong. Handle function have to take 2 params:
1. Dereferenced datascript database
2. Signal vector
This function should return transaction.

In your case, I would recommend using`reg-event-fx` from `re-frame` library together with `:ds` coeffect
```clojure
(ns example.events
   (:require [re-frame.core :as r]))

(r/reg-event-fx
   :my-event
   [(r/inject-cofx :ds)] ;; inject coeffect
   (fn [{:keys [ds]} [_ signal params]] ;; ds here is the DataScript database
      (GET
        "/upload"
        :format :transit
        :response-format :transit
        :handler #(re-frame/dispatch  [:create-new-db  %1])
        :error-handler #(js/alert (str "Didn't Work !" " Response:" ds)))
    ds)))
```

Hope it'll help

Regards,
Denis

**Comment by zachcp** [comment]
Denis,

Thanks for your quick response. 

So it looks like I had two sources of error - a misunderstanding of the re-frame/re-posh model AND an issue with transit serialization of datascript DBs.  After refactoring to use your format I end up with a cascade of errors ending in the same JSON error (see below).  But I should be able to try a simpler AJAX request for something like an integer to see if I can get the re-posh cascade correct. Then try to tackle the dastascript encoding issue. ( If you have any insight on that angle, I would also love to hear it.)

<img width="975" alt="screen shot 2017-12-29 at 4 57 44 pm" src="https://user-images.githubusercontent.com/1260667/34448243-8def6fa6-ecb9-11e7-901f-9667e826098c.png">

Thanks!
zach cp

**Comment by denistakeda** [comment]
@zachcp yeah, you are right, looks like this is a parsing problem. 

As far as I know, datascript uses `#datascript` extension for edn. To use this extension you have to have the proper parser on both, backend and frontend. Try to use this library for serialization/deserialization: https://github.com/tonsky/datascript-transit

If it won't help, could you please create the minimal repo where I can reproduce the error and play with it a little bit

**Comment by zachcp** [comment]
@denistakeda I've added an example to the todomvc that should provide a minimal set of features of what I was trying to do.  In that PR I've used your suggestions on how to make the GET call and have simplified the communication by:

1.  using EDN for serialization (handled easily on server and client with no extra work)
2. providing a list of datoms as per your initial-DB rather than providing a full datascript DB.

Once again, thank you for your library and for your prompt feedback.

zach cp
